Link summary related

Dorus opened the pull request:
A list of nice-to-have things:
* [x] A changelog 
* [ ] The link from the PR to the issue that it fixes. -> Not a direct fix, but this is work towards better link visualization to aid #409, #414
* [x] A description of what testing was done.
I clicked around on the new unrelated links and you navigate to new links from there.

As a followup to #408 this is an implementation of #414 (Only the visualization part):

![afbeelding](https://github.com/user-attachments/assets/bef76af3-ded1-4f2c-b290-dca8d76346f2)

As seen above, the link summary now not only display the currently selected link, but also recipes from other links covering the same recipe. It also tells you where in the recipe tree those links are, parent means they're in the same branch but higher up, child means the same but lower, unrelated is any other recipe.

There are a few more unimplemented ideas in #414 to add buttons to manipulate the other recipes, those can be tackled later.

![afbeelding](https://github.com/user-attachments/assets/2a2b592b-b68a-4b8a-b26e-667113464c79)

DaleStan commented:
A couple thoughts from looking at this:
* The new header says "Related links", but it displays related recipes instead. I'm not sure which you intended, but they should match. (Also, the new header should be drawn with `(TextBlockDisplayStyle)Font.subheader` or `Font.productionTableHeader`)
* Clicking on a related linked recipe didn't do what I expected, but now that I understand what it's doing, I think it's correct.

Dorus commented:
Thanks for the feedback! That's very useful.

"Related links" refers to the fact those recipes are part of related links (aka links that link the same item). I'll continue to think of a better name. "Items with related links" is a mouthful, but might be more clear. I'm also considering to split related links into 3 categories, one for child links, one for parent links and one for sibling links.

Dorus commented:
This PR is now ready to be merged:
- I've updated the description in the first post to match the current state.
- I've updated the remarks made by DaleStan
- A new full review would be appreciated.

veger commented:
Sorry I did not chime in earlier, I didn't have the time immediately and lost track of the notification.

I would make the new headings Parent/Child/etc links smaller. Now they are the same size as Production and Consumption and look equally important. But instead that are 'just' additional info on (similar) linked items elsewhere.

Or some horizontal line, or empty space between. Just to differentiate.

veger commented on Yafc/Workspace/ProductionTable/ProductionLinkSummaryScreen.cs:
```diff
@@ -39,8 +39,92 @@ private void BuildScrollArea(ImGui gui) {
+            Dictionary<ProductionLink, List<(RecipeRow row, float flow)>> childLinks = [], parentLinks = [], otherLinks = [];
+            List<(RecipeRow row, float flow)> unlinked = [], table;
+            foreach (var (r, relationLinks) in
+                GetAllRecipes(page, link.owner)
+                .Select(e => (e, IsParrent(e, parents) ? parentLinks : otherLinks))
+                .Concat(GetAllRecipes(link.owner).Select(e => (e, childLinks)))) {
+                if (link.capturedRecipes.Any(e => e == r)
+                    || (!r.Ingredients.Any(e => e.Goods == link.goods)
+                    && !r.Products.Any(e => e.Goods == link.goods)
+                    && !(r.fuel is not null && r.fuel == link.goods))) {
+                    continue;
+                }
```
Can you explain (in a comment maybe) what you are doing here? I cannot make much of it with all the (ultra) compact checks and stuff... (I am not too used to this (C#) syntax)

Like "Iterate over XYZ" and "Check if recipe ..." (at least I guess `r` is for recipe)

veger replied:
I think a white line just in front of the `if` statement could help as well. (It took me 3 reads to notice that is separate of the `foreach` due to the unfortunate indentation of the the multi-line `foreach` and the indentation of the block containing the `if` statement...

Same with the `continue` statement with the same indentation of the multi-line `if` conditions...

This all makes it a big code blob, while they are basically 3 separate blocks/expressions.

Dorus replied:
The first 3 lines are part of the for, and retrieve all recipes, first all recipes not nested on the current line, then all nested recipes (childeren).
The next 4 lines are an if filter to ignore the current link (first line) and unrelated links (the next 3 lines).

I'll refactor that into some functions.

veger commented on Yafc/Workspace/ProductionTable/ProductionLinkSummaryScreen.cs:
```diff
@@ -39,8 +39,92 @@ private void BuildScrollArea(ImGui gui) {
+        }
+    }
+
+    private bool IsParrent(RecipeRow row, List<ModelObject> parents) => row.Ingredients.Select(e => e.Link).Concat(row.Products.Select(e => e.Link)).Append(row.FuelInformation.Link)
```
typo: `IsParrent` -> `IsParent`

veger commented on Yafc/Workspace/ProductionTable/ProductionLinkSummaryScreen.cs:
```diff
@@ -39,8 +39,92 @@ private void BuildScrollArea(ImGui gui) {
+        if (o is ProductionTable page) {
+            Dictionary<ProductionLink, List<(RecipeRow row, float flow)>> childLinks = [], parentLinks = [], otherLinks = [];
+            List<(RecipeRow row, float flow)> unlinked = [], table;
+            foreach (var (r, relationLinks) in
```
as mentioned elsewhere: `r` is for recipe? If so, I would rename. It took me a while to figure out, making the code less readable (for me at least).

Dorus replied:
actually it's `row`, but `recipe` also works. The type is `recipeRow` xD

veger commented on Yafc/Workspace/ProductionTable/ProductionLinkSummaryScreen.cs:
```diff
@@ -39,8 +39,92 @@ private void BuildScrollArea(ImGui gui) {
+        }
+    }
+
+    private bool IsParrent(RecipeRow row, List<ModelObject> parents) => row.Ingredients.Select(e => e.Link).Concat(row.Products.Select(e => e.Link)).Append(row.FuelInformation.Link)
```
Should this function be part of `RecipeRow`, so you can do `recipeRow->IsParent(parents)` instead?

Dorus replied:
Mmm, this is a quite specific `IsParent` function for finding parents of a link. It scans all ingredients, products and fuel and checks them for the `link` owner (the `link` isn't even a parameter, it's a local field).
I've renamed it to `isLinkParent` instead.

veger commented on Yafc/Workspace/ProductionTable/ProductionLinkSummaryScreen.cs:
```diff
@@ -39,8 +39,92 @@ private void BuildScrollArea(ImGui gui) {
+        .Where(e => e!.goods == link.goods)
+        .Any(e => parents.Contains(e!.ownerObject!));
+
+    private static List<RecipeRow> GetAllRecipes(ProductionTable p, ProductionTable? skip = null) => p == skip ? [] : [..p.recipes, ..p.recipes
```
Should this function be part of `ProductionTable`, so you can do `productionTable->GetAllRecipes(skip)` instead?

Dorus replied:
yeah possibly. There was one already that i did not want to mess with, but i probably should. I had to add the `skip` option.

Dorus replied:
I convinced myself i could mess with it and deleted the duplicated code.

veger replied:
Can you add the new `skip` option to its function documentation?

Dorus replied:
Yeah, based on the doc it should do a depth-first ordering. Made some more changes to get that property back, idk how important it was but didn't want to mess with it. Also updated the docs.

veger commented on Yafc/Workspace/ProductionTable/ProductionLinkSummaryScreen.cs:
```diff
@@ -64,7 +148,7 @@ private void DestroyLink(ProductionLink link) {
     protected override void ReturnPressed() => Close();
 
-    private void BuildFlow(ImGui gui, List<(RecipeRow row, float flow)> list, float total, bool isLinkOutput) {
+    private void BuildFlow(ImGui gui, List<(RecipeRow row, float flow)> list, float total, bool isLinkOutput, int c = 0) {
```
`c` -> `colorIndex` (it is more descriptive/readable)

veger commented on Yafc/Workspace/ProductionTable/ProductionLinkSummaryScreen.cs:
```diff
@@ -118,6 +205,17 @@ void changeLinkView(ProductionLink newLink) {
         }
     }
 
+    private static SchemeColor GetColor(int c) => (c % 7) switch {
```
I would rename to `GetFlowColor()` making it more clear that this is the list of colors meant for the 'flow' field. (or 'embed' in the `BuildFlow()` function)

veger commented on Yafc/Workspace/ProductionTable/ProductionLinkSummaryScreen.cs:
```diff
@@ -118,6 +205,17 @@ void changeLinkView(ProductionLink newLink) {
+        4 => SchemeColor.Grey,
+        5 => SchemeColor.TagColorRed,
+        6 => SchemeColor.TagColorYellow,
+        _ => throw new NotImplementedException(),
```
```suggestion
        _ => throw new UnreachableException(),
```

Dorus commented on Yafc/Workspace/ProductionTable/ProductionLinkSummaryScreen.cs:
```diff
@@ -39,8 +39,92 @@ private void BuildScrollArea(ImGui gui) {
+                }
+                else {
+                    table.Add((r, 0));
+                }
```
Oops, there used to be a reason for this (one side multiplied by -1), but that has been removed since. Fixed.

Dorus commented:
> Sorry I did not chime in earlier, I didn't have the time immediately and lost track of the notification.
> 
> I would make the new headings Parent/Child/etc links smaller. Now they are the same size as Production and Consumption and look equally important. But instead that are 'just' additional info on (similar) linked items elsewhere.
> 
> Or some horizontal line, or empty space between. Just to differentiate.

I did both, indeed looks better:
![afbeelding](https://github.com/user-attachments/assets/4cb57805-9481-4635-bee2-b6134cedb817)

veger reviewed (approved):
Nice, UI indeed is better now!

For me all looks fine now. But let's wait for the others.

shpaass reviewed (approved):
Thank you for the feature!

shpaass commented:
I'll squash the fixups today.

Dorus commented:
Ah yeah i remembered to use proper commit messages this time, but did forget about fixup again. I kept the commits separate to easier track what comments i worked on, but many of them can be squashed now. Seems there's a conflict so if i resolve that, i'll also squash some commits. (Or feel free to beat me to it ^_^)

shpaass commented:
I'll wait for your changes.

veger commented:
You have a typo in the 2nd commit message :wink:
